src/compiler/transformers/rewrite_aliased_paths.ts

```diff
@@ -0,0 +1,152 @@
+    };
+  };
+}
+
+/**
+ * Transform modules aliased with `paths` in `tsconfig.json` to relative
+ * imported in source files.
+ *
+ * @returns a TypeScript transformer factory
+ */
+export const rewriteAliasedSourceFilePaths = (): ts.TransformerFactory<ts.SourceFile> => {
+  return (transformCtx: ts.TransformationContext) => {
+    const compilerHost = ts.createCompilerHost(transformCtx.getCompilerOptions());
+
+    return (tsSourceFile) => {
+      return ts.visitEachChild(tsSourceFile, visit(compilerHost, transformCtx, tsSourceFile.fileName), transformCtx);
+    };
+  };
+};
+
+/**
+ * This visitor function will modify any {@link ts.ImportDeclaration} nodes to
+ * rewrite module identifiers which are configured using
+ * the `paths` parameter in `tsconfig.json` from whatever name they are bound to
+ * to a relative path from the importer to the importee.
+ *
+ * We need to handle this ourselves because while the TypeScript team supports
+ * using the `paths` configuration to allow location-independent imports across
+ * a project (i.e. importing a module without having to use its relative
+ * path from the importing module) the TypeScript compiler has no built-in
+ * support for resolving these identifiers to the actual modules they point to
+ * in the `.js` and `.d.ts` files that it emits.
+ *
+ * So, for instance, if you have this set in `paths`:
+ *
+ * ```json
+ * "paths": {
+ *   "@utils": ["src/utils/index.ts""],
+ * }
+ * ```
+ *
+ * Then you'll be able to import it anywhere in your project:
+ *
+ * ```ts
+ * // src/importing.ts
+ * import { myUtil } from '@utils';
+ * ```
+ *
+ * but unfortunately, in the compiled output you'll still have:
+ *
+ * ```js
+ * // dist/importing.js
+ * import { myUtil } from "@utils";
+ * ```
+ *
+ * instead of what you _most likely_ want, which is:
+ *
+ * ```js
+ * // dist/importing.js
+ * import { myUtil } from "./utils";
+ * ```
+ *
+ * The TypeScript team have stated pretty unequivocally that they will not
+ * automatically resolve these identifiers to relative paths in output code
+ * {@see https://github.com/microsoft/TypeScript/issues/10866} and have
+ * said that resolving these module identifiers is the responsability of module
+ * bundling and build tools.
+ *
+ * So that means we've got to do it!
+ *
+ * This visitor function does so by getting the resolved file path to any
+ * module which is not 1) not external (i.e. not a dependency) and 2) is not
+ * already a relative, file-path based import. It then replaces the module
+ * identifier with the relative path from the importer to the importee.
+ *
+ * @param compilerHost a TS compiler host
+ * @param transformCtx a TS transformation context
+ * @param sourceFilePath the path to the source file being visited
+ * @returns a visitor which takes a node and optionally transforms imports
+ */
+const visit = (compilerHost: ts.CompilerHost, transformCtx: ts.TransformationContext, sourceFilePath: string) => {
```

Comment:
this 'visitor' is closely based on the one found in https://github.com/ionic-team/stencil/blob/main/src/compiler/transformers/map-imports-to-path-aliases.ts. It's basically just be modified to not be dependent on output target-specific stuff.